src/GraphBLAS-sharp.Backend/CSRMatrix/Mxm/Expansion.fs

```diff
@@ -0,0 +1,475 @@
+                            ndRange
+                            matrixRight.RowPointers
+                            matrixLeft.Columns
+                            rowPtrsForCols
+                            size)
+            )
+            processor.Post(Msg.CreateRunMsg<_, _>(kernel))
+
+            rowPtrsForCols
+
+    let private updateRowPtrsForCols
+        (clContext: ClContext)
+        workGroupSize =
+
+        let update =
+            <@
+                fun (ndRange: Range1D)
+                    (firstRowPointers: ClArray<int>)
+                    (size: int)
+                    (firstColumns: ClArray<int>)
+                    (secondRowPointers: ClArray<int>)
+                    (rowPtrsForCols: ClArray<int>) ->
+
+                    let i = ndRange.GlobalID0
+
+                    if i < size then
+                        let index = firstRowPointers.[i]
+                        if index <> firstRowPointers.[i + 1] then
+                            rowPtrsForCols.[index] <- secondRowPointers.[firstColumns.[index]]
+            @>
+        let kernel = clContext.CreateClKernel(update)
+
+        fun (processor: MailboxProcessor<_>)
+            (matrixLeft: CSRMatrix<'a>)
+            (matrixRight: CSRMatrix<'a>)
+            (rowPtrsForCols: ClArray<int>) ->
+
+            let size = matrixLeft.RowCount
+
+            let ndRange = Range1D.CreateValid(size, workGroupSize)
+            processor.Post(
+                Msg.MsgSetArguments
+                    (fun () ->
+                        kernel.ArgumentsSetter
+                            ndRange
+                            matrixLeft.RowPointers
+                            size
+                            matrixLeft.Columns
+                            matrixRight.RowPointers
+                            rowPtrsForCols)
+            )
+            processor.Post(Msg.CreateRunMsg<_, _>(kernel))
+
+    let private getRowPtrsForCols
+        (clContext: ClContext)
+        workGroupSize =
+
+        let initRowPtrsForCols = initRowPtrsForCols clContext workGroupSize
+        let updateRowPtrsForCols = updateRowPtrsForCols clContext workGroupSize
+        let scanByHeadFlagsInclude = PrefixSum.byHeadFlagsInclude <@ (+) @> clContext workGroupSize
+
+        fun (processor: MailboxProcessor<_>)
+            (headFlags: ClArray<int>)
+            (matrixLeft: CSRMatrix<'a>)
+            (matrixRight: CSRMatrix<'a>) ->
+
+            let rowPtrsForCols = initRowPtrsForCols processor matrixLeft matrixRight
+            updateRowPtrsForCols processor matrixLeft matrixRight rowPtrsForCols
+            let resRowPtrsForCols = scanByHeadFlagsInclude processor headFlags rowPtrsForCols 0
+            processor.Post(Msg.CreateFreeMsg<_>(rowPtrsForCols))
+            resRowPtrsForCols
+
+    let private getHeadFlags
+        (clContext: ClContext)
+        workGroupSize =
+
+        let create = ClArray.create clContext workGroupSize
+
+        let init =
```

Comment:
There are at least three identical kernels: this one, [this](https://github.com/YaccConstructor/GraphBLAS-sharp/pull/35/files#diff-692e43c69d92aaf4f3640dbcdea156ecd278fef240f1ca54bf2cbbca76fd013dR135), and [this](https://github.com/YaccConstructor/GraphBLAS-sharp/pull/35/files#diff-692e43c69d92aaf4f3640dbcdea156ecd278fef240f1ca54bf2cbbca76fd013dR73)